placement/mappings.go

```diff
@@ -0,0 +1,232 @@
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package placement
+
+import (
+	"time"
+
+	"github.com/m3db/m3storage"
+	"github.com/m3db/m3storage/generated/proto/schema"
+	"github.com/m3db/m3x/log"
+	"github.com/m3db/m3x/time"
+
+	"github.com/willf/bitset"
+)
+
+// mappings maintains information about all of the shard mappings in a
+// database
+type mappings struct {
+	name   string
+	active []*activeMappingRule
+	log    xlog.Logger
+}
+
+func newMappings(name string, log xlog.Logger) *mappings {
+	return &mappings{
+		name: name,
+		log:  log,
+	}
+}
+
+// clone clones the mappings
+func (m *mappings) clone() *mappings {
+	clone := &mappings{
+		name:   m.name,
+		active: make([]*activeMappingRule, len(m.active)),
+		log:    m.log,
+	}
+
+	for n, a := range m.active {
+		clone.active[n] = a.clone()
+	}
+
+	return clone
+}
+
+// apply generates new mappings based on the given set of rules
+func (m *mappings) apply(rules *schema.ClusterMappingRuleSet, newClusterVersions map[string]int) error {
```

Comment:
It seems `apply` only sets the versions for clusters with initial shard assignments (i.e., shards with no `FromCluster`) as well as those in `rules.ClusterConfigUpdates`. How do we handle newly joined clusters with shards that were previously assigned to other clusters?

For example, let's say shard 0 to 3 were assigned to cluster A, now cluster B joins and takes ownership of 2 and 3. The rules in this case will indicate shard 2 and shard 3 have been transferred from cluster A to cluster B.

Therefore in the `apply` method, because shard 2 and shard 3 have a valid from cluster (i.e., cluster A), and AFAICT cluster B is not included in `rules.ClusterConfigUpdates`, it seems we are not setting the version for cluster B and therefore not including cluster B in the list of known clusters in the cluster mapping provider. Did I miss anything here?